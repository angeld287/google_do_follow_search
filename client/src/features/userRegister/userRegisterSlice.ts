import { createSlice, PayloadAction } from '@reduxjs/toolkit';
import { RootState } from '../../app/store';
import { registerAsync } from './asyncThunks';
import { IUserRegisterSlice } from './IUserRegister';

export const initialState: IUserRegisterSlice = {
  user: {
    id: "",
    email: "",
    phoneNumber: "",
    password: "",
    confirmPassword: "",
    fullname: "",
    gender: "",
    userName: "",
  },
  status: 'idle',
};

export const userRegisterSlice = createSlice({
  name: 'userRegister',
  initialState,

  //Actions
  reducers: {
    setEmail: (state, action: PayloadAction<string>) => {
      state.user.email = action.payload;
    },
  },

  //async operations
  extraReducers: (builder) => {
    builder
      .addCase(registerAsync.pending, (state) => {
        state.status = 'pending';
      })
      .addCase(registerAsync.fulfilled, (state, action) => {
        state.status = 'idle';
      })
      .addCase(registerAsync.rejected, (state) => {
        state.status = 'failed';
      })
  },
});

//Actions
export const { setEmail } = userRegisterSlice.actions;

export const selectUserRegister = (state: RootState) => state.userRegister;

export default userRegisterSlice.reducer;
