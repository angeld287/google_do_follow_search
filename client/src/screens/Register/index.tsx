import React, { useCallback, useState } from "react"
import CustomForm from "../../components/CustomForm"
import { Content } from 'antd/lib/layout/layout';
import styles from "./styles";
//import { useAppDispatch } from "../../app/hooks";
import { MessageApi } from "antd/lib/message";
import { IValidationError } from "../../interfaces/models/IBase";
import { ICustomInputGroup } from "../../components/CustomInputGroup/ICustomInputGroup";
import { ICustomButton } from "../../components/CustomButton/ICustomButton";


const Register: React.FC = () => {

    //const dispatch = useAppDispatch()
    const [message, setMessage] = useState<MessageApi>()
    const [error, setError] = useState<IValidationError>()

    const handleClicRegisterButton = useCallback(
        (message: MessageApi) => {
            setMessage(message);
        }
        , []
    );

    let inputFields: Array<ICustomInputGroup> = [
        {
            name: 'email',
            label: 'Email',
            defaultValue: '',
            disabled: false,
        },
        {
            name: 'username',
            label: 'Username',
            defaultValue: '',
            disabled: false,
        },
        {
            name: 'phonenumber',
            label: 'Phone Number',
            defaultValue: '',
            disabled: false,
        },
        {
            name: 'password',
            label: 'Password',
            defaultValue: '',
            disabled: false,
            type: 'password'
        },
        {
            name: 'confirmpassword',
            label: 'Confirm your password',
            defaultValue: '',
            disabled: false,
            type: 'password'
        },
        {
            name: 'fullname',
            label: 'Fullname',
            defaultValue: '',
            disabled: false,
        },
        {
            name: 'gender',
            label: 'Gender',
            defaultValue: '',
            disabled: false,
        }
    ]

    let btns: Array<ICustomButton> = [
        {
            color: 'blue',
            _key: 'register_btn',
            children: 'Sign Up',
            loading: false,
            htmlType: 'submit',
            name: "register",
        }
    ]

    return (
        <>
            <Content style={styles.container}>
                <CustomForm error={error} onSubmit={handleClicRegisterButton} fields={inputFields} buttons={btns} verticalButtons={false} loading={false} />
            </Content>
        </>
    )
}

export default Register