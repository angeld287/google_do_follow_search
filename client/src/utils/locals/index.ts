// @ts-nocheck
/**
 * Define App Locals & Configs
 *
 * @author Angel Angeles <aangeles@litystyles.com>
 */

class Locals {

    /**
     * Makes env configs available for your app
     * throughout the app's runtime
     */
    public static config(): any {
        const server_host = "52.70.36.16";
        //(
        //    window._env_ ?
        //        window._env_.REACT_APP_SERVER_HOST : (process.env.SERVER_HOST || "localhost"));

        const server_port = "3000"
        //(
        //    window._env_ ?
        //        window._env_.REACT_APP_SERVER_PORT : (process.env.REACT_APP_SERVER_PORT || 3000));

        const server_url = "http://" + server_host + ":" + server_port

        const a_tags_reg_exp = /(<a\s*(?!.*\bnofollow)[^>]*)(href="https?:\/\/)((?!(?:(?:www\.)?'.implode('|(?:www\.)?', $follow_list).'))[^"]+)"((?!.*\bnofollow)[^>]*)(?:[^>]*)>/g
        const follow_reg_exp = /rel=("|')dofollow("|')/g
        const no_follow_reg_exp = /rel=("|')nofollow("|')/g

        return {
            server_url,
            a_tags_reg_exp,
            follow_reg_exp,
            no_follow_reg_exp
        }
    }
}

export default Locals;