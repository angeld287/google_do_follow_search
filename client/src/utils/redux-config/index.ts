import { configureStore } from "@reduxjs/toolkit";
import userSessionReducer from '../../features/userSession/userSessionSlice';
import userRegisterReducer from '../../features/userRegister/userRegisterSlice';
import userSearchSlice from '../../features/googleSearch/searchSlice';

export const store = configureStore({
    reducer: {
        userSession: userSessionReducer,
        userRegister: userRegisterReducer,
        googleSearch: userSearchSlice,

    },
    middleware: (getDefaultMiddleware) => getDefaultMiddleware({
        serializableCheck: false
    })
});