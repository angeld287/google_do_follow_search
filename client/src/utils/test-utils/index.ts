// test-utils.js
import { render, queries, RenderOptions, RenderResult } from '@testing-library/react'
import { JSXElementConstructor, ReactElement } from 'react';
import * as customQueries from './custom-queries'
import * as customFunctionalities from './functionalities';

const customRender = (ui: ReactElement<any, string | JSXElementConstructor<any>>, options?: RenderOptions) =>
    render(ui, { queries: { ...queries, ...customQueries, ...customFunctionalities }, ...options })

// re-export everything
export * from '@testing-library/react'

// override render method
export { customRender as render }