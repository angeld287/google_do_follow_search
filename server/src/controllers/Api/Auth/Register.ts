/**
 * Define the Register API logic
 *
 * @author Angel Angeles <aangeles@litystyles.com>
 */

import { validationResult } from 'express-validator';
import IUserService from '../../../interfaces/IUserService';
import { IRequest, IResponse } from '../../../interfaces/vendors';
import Log from '../../../middlewares/Log';
import userService from '../../../services/userService';

class Register {
    /**
     * Execute the action of register an user if the inputs are valid
     * @param {string} req: get the request from the post
     * @param {string} res: the response expected by the post
     * @return {Promise<>} return a promise with the json result
     */
    public static async perform(req: IRequest, res: IResponse): Promise<any> {
        try {
            const errors = validationResult(req);
            let user: IUserService = new userService()

            if (!errors.isEmpty()) {
                return res.json({
                    errors: errors.array()
                });
            }

            const _email = req.body.email;
            const _phoneNumber = req.body.phoneNumber;
            const _userName = req.body.username;
            const _password = req.body.password;
            const _fullName = req.body.fullName;
            const _gender = req.body.gender;

            const existenceVerifications = await Promise.all(
                [
                    user.verifyIfEmailExist(_email),
                    user.verifyIfPhoneNumberExist(_phoneNumber),
                    user.verifyIfUserNameExist(_userName)
                ]
            );

            if (existenceVerifications.filter(_ => _.exist).length > 0) {
                return res.json({
                    errors: existenceVerifications.filter(_ => _.exist)
                });
            }

            const createUser = await user.createNewUser(_email, _phoneNumber, _password, _fullName, _gender, _userName, null);

            if (createUser.created) {
                Log.info(`New user created ` + _userName);
                return res.json({
                    userId: createUser.id,
                    message: 'The user has been created successfully'
                });
            } else {
                Log.error(`An error was occurred while creating the user`);
                return res.status(500).json({
                    error: 'An error was occurred while creating the user',
                });
            }

        } catch (error) {
            Log.error(`Internal Server Error ` + error);
            return res.status(500).json({
                error: 'Internal Server Error',
            });
        }
    }

    /**
     * Verify if the field password match with the confirmPassword
     * @param {string} value: the values of the field password
     * @param {string} req: get the request from the post
     * @return {string} return the values of the field password
     */
    public static verifyPasswordsMatch(value, req) {
        if (value !== req.body.confirmPassword) {
            // trow error if passwords do not match
            throw new Error("Passwords don't match.");
        } else {
            return value;
        }
    }

    /**
     * Verify if the field phoneNumber match with the correct format:
     * Valid formats: (123) 456-7890, (123)456-7890, 123-456-7890, 123.456.7890, 1234567890, +31636363634, 075-63546725
     * @param {string} value: the values of the field phoneNumber
     * @return {string} return a boolean with the validation of the phone number format
     */
    public static isPhoneNumber(value) {
        try {
            return value.match(/^[\+]?[(]?[0-9]{3}[)]?[-\s\.]?[0-9]{3}[-\s\.]?[0-9]{4,6}$/im);
        } catch (error) {
            throw new Error(error);
        }
    }
}

export default Register;
