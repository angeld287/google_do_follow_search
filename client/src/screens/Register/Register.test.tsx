import { configureStore } from '@reduxjs/toolkit';
import React from 'react';
import { Provider } from 'react-redux';
import { userRegisterSlice } from '../../features/userRegister/userRegisterSlice';
import { store } from '../../utils/redux-config';
import { render, RenderResult, functions, fireEvent, screen, waitFor, act } from '../../utils/test-utils';
import Register from './index'

let component: RenderResult;

describe("Login Test Suite", () => {

    beforeEach(() => {
        if (component)
            component.unmount()

        act(() => {
            component = render(
                <Provider store={store}>
                    <Register />
                </Provider>
            )
        });
    });

    test('It must render the component correcly', () => {
        expect(component.asFragment()).toMatchSnapshot();
    })

    test('It must display basic register components', () => {
        expect(component.getByText(/Email/i)).toBeInTheDocument();
        expect(component.getByText(/Username/i)).toBeInTheDocument();
        expect(component.getByText(/Phone Number/i)).toBeInTheDocument();
        expect(component.getAllByText(/Password/i)[0]).toBeInTheDocument();
        expect(component.getByText(/Confirm your password/i)).toBeInTheDocument();
        expect(component.getByText(/Fullname/i)).toBeInTheDocument();
        expect(component.getByText(/Gender/i)).toBeInTheDocument();

        expect(component.getByText(/Sign Up/i)).toBeInTheDocument();
    });

    test('It must respond "E-mail cannot be blank." when email is blank.', async () => {

        await act(() => {
            fireEvent.click(screen.getByText(/Sign Up/i));
        });

        await waitFor(() => {
            expect(screen.getByText("E-mail cannot be blank.")).toBeInTheDocument();
        });
    });

    test('It should respond "Username cannot be blank." when Username is blank.', async () => {

        functions.writeInInputFoundByPlaceHolder(null, /Email/i, "existingadmin@test.com");

        await act(() => {
            fireEvent.click(screen.getByText(/Sign Up/i));
        });

        await waitFor(() => {
            expect(screen.getByText("Username cannot be blank.")).toBeInTheDocument();
        });
    });
});