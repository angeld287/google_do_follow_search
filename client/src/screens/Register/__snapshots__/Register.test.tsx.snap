// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login Test Suite It must render the component correcly 1`] = `
<DocumentFragment>
  <main
    class="ant-layout-content"
    style="width: 400px; height: 400px; margin: 50px auto; background-color: white; flex-direction: column; justify-content: center; align-items: center;"
  >
    <form
      autocomplete="off"
      class="ant-form ant-form-horizontal"
      id="basic"
    >
      <div>
        <div
          class="ant-row ant-form-item"
        >
          <div
            class="ant-col ant-col-8 ant-form-item-label"
          >
            <label
              class=""
              for="basic_email"
              title="Email"
            >
              Email
            </label>
          </div>
          <div
            class="ant-col ant-col-16 ant-form-item-control"
          >
            <div
              class="ant-form-item-control-input"
            >
              <div
                class="ant-form-item-control-input-content"
              >
                <input
                  class="ant-input"
                  id="basic_email"
                  placeholder="Email"
                  type="text"
                  value=""
                />
              </div>
            </div>
          </div>
        </div>
      </div>
      <div>
        <div
          class="ant-row ant-form-item"
        >
          <div
            class="ant-col ant-col-8 ant-form-item-label"
          >
            <label
              class=""
              for="basic_username"
              title="Username"
            >
              Username
            </label>
          </div>
          <div
            class="ant-col ant-col-16 ant-form-item-control"
          >
            <div
              class="ant-form-item-control-input"
            >
              <div
                class="ant-form-item-control-input-content"
              >
                <input
                  class="ant-input"
                  id="basic_username"
                  placeholder="Username"
                  type="text"
                  value=""
                />
              </div>
            </div>
          </div>
        </div>
      </div>
      <div>
        <div
          class="ant-row ant-form-item"
        >
          <div
            class="ant-col ant-col-8 ant-form-item-label"
          >
            <label
              class=""
              for="basic_phonenumber"
              title="Phone Number"
            >
              Phone Number
            </label>
          </div>
          <div
            class="ant-col ant-col-16 ant-form-item-control"
          >
            <div
              class="ant-form-item-control-input"
            >
              <div
                class="ant-form-item-control-input-content"
              >
                <input
                  class="ant-input"
                  id="basic_phonenumber"
                  placeholder="Phone Number"
                  type="text"
                  value=""
                />
              </div>
            </div>
          </div>
        </div>
      </div>
      <div>
        <div
          class="ant-row ant-form-item"
        >
          <div
            class="ant-col ant-col-8 ant-form-item-label"
          >
            <label
              class=""
              for="basic_password"
              title="Password"
            >
              Password
            </label>
          </div>
          <div
            class="ant-col ant-col-16 ant-form-item-control"
          >
            <div
              class="ant-form-item-control-input"
            >
              <div
                class="ant-form-item-control-input-content"
              >
                <input
                  class="ant-input"
                  id="basic_password"
                  placeholder="Password"
                  type="password"
                  value=""
                />
              </div>
            </div>
          </div>
        </div>
      </div>
      <div>
        <div
          class="ant-row ant-form-item"
        >
          <div
            class="ant-col ant-col-8 ant-form-item-label"
          >
            <label
              class=""
              for="basic_confirmpassword"
              title="Confirm your password"
            >
              Confirm your password
            </label>
          </div>
          <div
            class="ant-col ant-col-16 ant-form-item-control"
          >
            <div
              class="ant-form-item-control-input"
            >
              <div
                class="ant-form-item-control-input-content"
              >
                <input
                  class="ant-input"
                  id="basic_confirmpassword"
                  placeholder="Confirm your password"
                  type="password"
                  value=""
                />
              </div>
            </div>
          </div>
        </div>
      </div>
      <div>
        <div
          class="ant-row ant-form-item"
        >
          <div
            class="ant-col ant-col-8 ant-form-item-label"
          >
            <label
              class=""
              for="basic_fullname"
              title="Fullname"
            >
              Fullname
            </label>
          </div>
          <div
            class="ant-col ant-col-16 ant-form-item-control"
          >
            <div
              class="ant-form-item-control-input"
            >
              <div
                class="ant-form-item-control-input-content"
              >
                <input
                  class="ant-input"
                  id="basic_fullname"
                  placeholder="Fullname"
                  type="text"
                  value=""
                />
              </div>
            </div>
          </div>
        </div>
      </div>
      <div>
        <div
          class="ant-row ant-form-item"
        >
          <div
            class="ant-col ant-col-8 ant-form-item-label"
          >
            <label
              class=""
              for="basic_gender"
              title="Gender"
            >
              Gender
            </label>
          </div>
          <div
            class="ant-col ant-col-16 ant-form-item-control"
          >
            <div
              class="ant-form-item-control-input"
            >
              <div
                class="ant-form-item-control-input-content"
              >
                <input
                  class="ant-input"
                  id="basic_gender"
                  placeholder="Gender"
                  type="text"
                  value=""
                />
              </div>
            </div>
          </div>
        </div>
      </div>
      <div
        style="text-align: right;"
      >
        <div
          class="ant-row ant-form-item"
        >
          <div
            class="ant-col ant-col-16 ant-col-offset-8 ant-form-item-control"
          >
            <div
              class="ant-form-item-control-input"
            >
              <div
                class="ant-form-item-control-input-content"
              >
                <button
                  class="ant-btn ant-btn-default"
                  style="color: blue;"
                  type="button"
                >
                  <span>
                    Login
                  </span>
                </button>
                <button
                  class="ant-btn ant-btn-default"
                  style="color: blue;"
                  type="submit"
                >
                  <span>
                    Sign Up
                  </span>
                </button>
              </div>
            </div>
          </div>
        </div>
      </div>
    </form>
  </main>
</DocumentFragment>
`;
